### PR TITLE
Feat/244 properly dispose reactive list contents

### PR DESCRIPTION
**Pull request by lxatrhea**
- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-IME [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-IME-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
Created DisposableReactiveList<IDisposable> that forces a developer to think about disposal of items in its underlying list.
All ReactiveList<SomeViewModel> in the solution should now be a DisposableReactiveList.

### CODE REVIEW

File: CDP4Composition/Mvvm/Types/DisposableReactiveList.cs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        }
+
+        /// <summary>
+        /// Overrides the method implementation in <see cref="ReactiveList{T}"/> and adds an <see cref="ObsoleteAttribute"/>
+        /// so the code doesn't compile anymore when the method is unexpectedly being used.
+        /// </summary>
+        [Obsolete("RemoveAll(IEnumerable<T> items) is deprecated, please use RemoveAll(IEnumerable<T> items, bool dispose) instead.", true)]
+        public new void RemoveAll(IEnumerable<T> items)
+        {
+            throw new NotImplementedException();
```
Comment (samatrhea) [inline CDP4Composition/Mvvm/Types/DisposableReactiveList.cs]: i think this should be a `NotSupportedException`

File: CDP4Composition/Mvvm/Types/DisposableReactiveList.cs [hunk 2]
```diff
@@ -0,0 +1,261 @@
+﻿// ------------------------------------------------------------------------------------------------
+// <copyright file="DisposableReactiveList.cs" company="RHEA System S.A.">
```
Comment (samatrhea) [inline CDP4Composition/Mvvm/Types/DisposableReactiveList.cs]: There is no corresponding TestFixture for the `DisposableReactiveList`, please add it and also write a page on the IME wiki explaining the use and why this class is so important.

Please also explain when the `DisposableReactiveList` shall be used and when the regular `ReactiveList` shall be used.

File: CDP4Composition/Mvvm/Types/DisposableReactiveList.cs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        }
+
+        /// <summary>
+        /// Overrides the method implementation in <see cref="ReactiveList{T}"/> and adds an <see cref="ObsoleteAttribute"/>
+        /// so the code doesn't compile anymore when the method is unexpectedly being used.
+        /// </summary>
+        [Obsolete("RemoveAt(int index) is deprecated, please use RemoveAt(int index, bool dispose) instead.", true)]
+        public new void RemoveAt(int index)
+        {
+            throw new NotImplementedException();
```
Comment (samatrhea) [inline CDP4Composition/Mvvm/Types/DisposableReactiveList.cs]: i think this should be a `NotSupportedException`

File: CDP4Composition/Mvvm/Types/DisposableReactiveList.cs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        }
+
+        /// <summary>
+        /// Overrides the method implementation in <see cref="ReactiveList{T}"/> and adds an <see cref="ObsoleteAttribute"/>
+        /// so the code doesn't compile anymore when the method is unexpectedly being used.
+        /// </summary>
+        [Obsolete("Remove(T item) is deprecated, please use  Remove(T item, bool dispose) instead.", true)]
+        public new bool Remove(T item)
+        {
+            throw new NotImplementedException();
```
Comment (samatrhea) [inline CDP4Composition/Mvvm/Types/DisposableReactiveList.cs]: i think this should be a `NotSupportedException`

File: CDP4Composition/Mvvm/Types/DisposableReactiveList.cs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        }
+
+        /// <summary>
+        /// Overrides the method implementation in <see cref="ReactiveList{T}"/> and adds an <see cref="ObsoleteAttribute"/>
+        /// so the code doesn't compile anymore when the method is unexpectedly being used.
+        /// </summary>
+        [Obsolete("RemoveRange(int index, int count) is deprecated, please use  RemoveRange(int index, int count, bool dispose) instead.", true)]
+        public new void RemoveRange(int index, int count)
+        {
+            throw new NotImplementedException();
```
Comment (samatrhea) [inline CDP4Composition/Mvvm/Types/DisposableReactiveList.cs]: i think this should be a `NotSupportedException`

File: CDP4Composition/Mvvm/Types/DisposableReactiveList.cs [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    public class DisposableReactiveList<T> : ReactiveList<T> where T : IDisposable
+    {
+        /// <summary>
+        /// Overrides the method implementation in <see cref="ReactiveList{T}"/> and adds an <see cref="ObsoleteAttribute"/>
+        /// so the code doesn't compile anymore when the method is unexpectedly being used.
+        /// </summary>
+        [Obsolete("Clear is deprecated, please use Clear(bool dispose) instead.", true)]
+        public new void Clear()
+        {
+            throw new NotImplementedException();
```
[the remark this answers is not in the document]
Comment (lxatrhea) [reply CDP4Composition/Mvvm/Types/DisposableReactiveList.cs]: GC.SuppressFinalize is not needed, so i'll remove it. 
In the end I did not add a Finalizer, so suppressing it doesn't do anything. 
We are going to need a finalizer when unmanaged resources need to be released, but for now we don't.

Typical use of IDisposable:

`    private bool disposed = false;

    protected virtual void Dispose(bool disposing)
    {
        if (!disposed)
        {
            if (disposing)
            {
                // called via myClass.Dispose(). 
                // OK to use any private object references
            }
            // Release unmanaged resources.
            // Set large fields to null.                
            disposed = true;
        }
    }

    public void Dispose() // Implement IDisposable
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    ~MyClass() // the finalizer
    {
        Dispose(false);
    }`

File: CDP4Composition/Mvvm/Types/DisposableReactiveList.cs (hunk 1 above)
Comment (alexatrhea) [reply CDP4Composition/Mvvm/Types/DisposableReactiveList.cs]: I concur.

**Comment by lxatrhea** [comment]
I'll remove the GC.SuppressFinalize, because there isn't even a finalizer there, so there is nothing to suppress.

**Comment by lxatrhea** [comment]
Also see: https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/wiki/MVVM#disposablereactivelistt for documentation.

**Pull request merged**